opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -43,22 +90,39 @@ impl<T: Number<T>> ValueMap<T> {
             self.no_attribute_value.add(measurement);
             self.has_no_value_attribute_value
                 .store(true, Ordering::Release);
-        } else if let Ok(mut values) = self.values.lock() {
-            let size = values.len();
-            match values.entry(attrs) {
-                Entry::Occupied(mut occupied_entry) => {
-                    let sum = occupied_entry.get_mut();
-                    *sum += measurement;
+        } else {
+            let bucket_index = Self::hash_to_bucket(&attrs) as usize;
+            let bucket_mutex = &self.buckets[bucket_index];
+
+            let mut bucket_guard = match bucket_mutex.lock() {
+                Ok(guard) => guard,
+                Err(e) => {
+                    eprintln!("Failed to acquire lock due to: {:?}", e);
+                    return; // TBD - retry ?
```

Comment:
In this failure path, does the measurement just get dropped? The only reason why a mutex lock would fail is if the mutex was poisoned. Reading the [documentation on mutex poisoning](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning)

> Once a mutex is poisoned, all other threads are unable to access the data by default as it is likely tainted (some invariant is not being upheld).

We should probably account for this error state and try to recover from the failure. A mutex is usually poisoned because another thread panicked while holding the lock. Since the underlying data this mutex is protecting is a map of measurements, I don't think there are external invariants that would be affected if we re-acquired the guard from within the mutex. This can be done using the [`into_inner`](https://doc.rust-lang.org/std/sync/struct.PoisonError.html#method.into_inner) method on the `PoisonError` type. This would get us access to the underlying data to update the measurement if we want to, but we would still need to re-instantiate the bucket with a new mutex.

Since `into_inner` gives us access to the inner hashmap, we can probably `mem::replace` the specific bucket with a fresh mutex with the existing data in the error path. 

This would definitely be a performance hit, but I think it's worth it because of the data integrity it affords.

Reply:
Thanks @bIgBV. This is the right approach. Seems we are not handling lock even in the earlier single-level hashmap. I will create a separate issue to handle this, to be fixed in a separate PR.

Reply:
created #1594 to handle mutex poisoning.